New ertp guide (#287)
Agoric/documentation: main/ertp/guide/PursesAndPayments.md

# Purses and Payments
 
Digital assets can be any of:
- Currency-like, such as our imaginary Quatloos currency.
- Goods-like digital assets, such as magic weapons for use in a game or theater tickets.
- Other kinds of rights, such as the right to participate in a particular contract.

In ERTP, digital assets exist in either a `purse` or a `payment`:
- **[`purse`](https://agoric.com/documentation/glossary/#purse)**: Holds

>   a quantity of same-branded digital assets until part or

Reviewer: Why introduce the word quantity when we already have amount?

  all of them are withdrawn into a `payment`. A new `purse` is created
  by an `issuer` and can only hold assets of that `issuer`'s `brand`. 
- **[`payment`](https://agoric.com/documentation/glossary/#payment)**:
  Holds a quantity of same-branded digital assets to transfer to another party. 
  A new `payment` is created either with new assets by a `mint` or by 
  withdrawing assets from a `purse`. It can only hold assets of the same `brand` as

>   that `mint` or `purse`. .

Reviewer: ```suggestion
  that `mint` or `purse`.
```

For any `brand`, any number of `purses` or `payments` can hold assets
of that `brand`. Neither a `purse` nor a `payment` can ever change their
associated `brand`.

> Each non-empty `purse` and `payment` object contains exactly one
> quantity of its digital asset. 

Reviewer: I think this is a very confusing statement. What does one quantity even mean?  I think we should replace this statement with:

```suggestion
Each `purse` and `payment` object contains a specific amount of digital assets, which may be none at all (`empty` in amountMath terms).
```

> For example, a single `purse` can contain assets of
> 5 Quatloos, but not two separated assets of 2 Quatloos and 3 Quatloos. To
> hold multiple, separated, quantities of same-branded assets you have to have multiple
> `purses`. This is similar to how, in the real world, you might have two bank accounts 
> for dollars. One is for everyday expenses, and the other is an emergency fund.

Reviewer: This explanation seems confusing. Can we say instead:

```suggestion
In the same way that you might have separate bank accounts for different purposes, you can make different purses for the same brand of digital asset. One purse might hold 2 Quatloos and another purse might hold 3 Quatloos.
```

fungible assets means it has a value of 0. For non-fungible 
assets, such as theatre tickets, it doesn't have any tickets.

Unlike `payments`, `purses` are not meant to be sent to others. To transfer 
digital assets, you should withdraw a `payment` from a `purse` and send 
the `payment` to another party.

You can create a *deposit facet* for a `purse`. Deposit facets are either sent to other parties or made publicly known. Any party can deposit a `payment` into the 
deposit facet, which deposits it into its associated `purse`. However, no one can
use a deposit facet to either make a withdrawal from its `purse` or get the `purse`'s balance. 

If you have a deposit facet, you make a deposit to its associated `purse` by calling 
`depositFacet.receive(payment);`. Note that you add a `payment` to a `purse` with a `deposit()` method, while you add a `payment` to a `depositFacet` with a `receive()` method. 

The `payment` must be the same `brand` as what
the associated `purse` object can contain. Otherwise it throws an error. 
When sending a party a 
deposit facet object, you should tell them what assets `brand` it accepts.

> ![Purse methods](./assets/purse.svg)  

Reviewer: The `(digital assets)` seems to be defining a purse, but a purse contains digital assets. I think digital assets should be an object in the diagram, connected by the line "can only contain". And Brand should go off digital assets, not the purse.

> Purses have four API methods. The following is a brief description and example of each `purse` method. For

Reviewer: Please remove the counts of methods

more detail, click the method's name to go to its entry in the [ERTP
API Reference](https://agoric.com/documentation/ertp/api/#ertp-api).

> - [`purse.getCurrentAmount()`](https://agoric.com/documentation/ertp/api/purse.html#purse-getcurrentamount)

Reviewer: relative reference please

     // that will be deposited:
     depositOnlyFacet.receive(payment);
     ```

In addition, the method to create a new, empty, `purse` is called on an `issuer`:
- [`issuer.makeEmptyPurse()`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-makeemptypurse)
  - Returns a new empty `purse` for storing digital assets of the `brand` the `issuer` is associated with.
  - ```js
    // The new purse can only contain assets of the Quatloos brand.
    const quatloosPurse = quatloosIssuer.makeEmptyPurse();
    ```
## Payments

![Payment methods](./assets/payment.svg)   

Payments hold digital assets intended to be transferred to another party.
They are linear, meaning that either a `payment` has its full
original balance, or it is used up entirely. It is impossible to
partially use a `payment`. 

In other words, if you create a `payment` containing
10 Quatloos, the `payment` will always either contain 
10 Quatloos or it will be deleted from its `issuer` records and no
longer have any value. While a `payment` can be either combined with others or
split into multiple `payments`, in both cases the original `payment(s)`

> is/are burned and the results put in one or more new `payments`.

Reviewer: ```suggestion
is/are deleted and the results put in one or more new `payments`.
```
Not burned. Let's reserve the word burned for when digital assets are destroyed.

A `payment` can be deposited in purses, split into multiple 
`payments`, combined with other `payments`, and claimed (getting an exclusive `payment` and revoking access from anyone else). 

A `payment` is often received from other parties. Since they are not
self-verifying, you cannot trust `payments`. To get the verified balance
of a `payment`, use the `getAmountOf(payment)` method on the trusted `issuer` 
for the `payment`'s `brand`. 

To get the `issuer` for a `brand` you didn't create, 
ask someone you trust. For example, the venue creating tickets for shows
can be trusted to give you the tickets' `issuer`. Or, a friend might have 
a cryptocurrency they like, and, if you trust them, you might accept 
that the `issuer` they give you is valid.

To convert a `payment` into a new `purse`:
1. Get the `payment`'s trusted `issuer`.
2. Use the `issuer` to create an empty `purse` for that `brand`.

> 3. Deposit the `payment` into the new `purse`. `purse.deposit(payment)` burns the `payment`.

Reviewer: ```suggestion
3. Deposit the `payment` into the new `purse`. `purse.deposit(payment)` deletes the `payment`.
```

`Payments` have only one API method, but many methods for other ERTP components
have `payments` as arguments and effectively operate on a `payment`. The following is a 
brief description and example of each `payment`-related method. For
more detail, click the method's name to go to its entry in the [ERTP
API Reference](https://agoric.com/documentation/ertp/api/#ertp-api).
- [`payment.getAllegedBrand()`](https://agoric.com/documentation/ertp/api/payment.html#payment-getallegedbrand)
  - Returns a `brand`, indicating what kind of digital asset the
  `payment` purports to be. Since a `payment` is not trusted, this result should be   
   treated with suspicion. Either verify the value before holding on to it, or check 

>    the result when you use it. Any successful operation by 

Reviewer: "verify the value before holding on to it" didn't make sense to me.

   the `issuer` for that `brand` done on the `payment` verifies the `payment`.

Other objects' `payment`-related methods:

- [`issuer.getAmountOf(payment)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-getamountof-payment)

>   - Get a description of a `payment` balance as an `amount`. The `payment` itself is not trusted,

Reviewer: Maybe say something like "get the amount of digital assets in the payment"

    so you must use the `issuer` method associated with its `brand` to be sure of getting the
    true value. 
    ```js
    const quatloosPayment = quatloosMint.mintPayment(quatloosAmountMath.make(100));
    quatloosIssuer.getAmountOf(quatloosPayment); // returns an amount with 100 value `quatloos` brand
    ```
- [`issuer.burn(payment, optAmount)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-burn-payment-optamount)
  - Burn all of the digital assets in the `payment`. `optAmount` is
    optional but if present, the `payment` balance must be equal to
    it. If `payment` is a promise, the operation proceeds after it resolves. 
  - ```js
    const amountToBurn = quatloosAmountMath.make(10);
    const paymentToBurn = quatloosMint.mintPayment(amountToBurn);
    // burntAmount should equal 10
    const burntAmount = quatloosIssuer.burn(paymentToBurn, amountToBurn);
    ```
- [`issuer.claim(payment, optAmount)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-claim-payment-optamount)
  - Transfer all assets from the `payment` to a returned new `payment`

>     and burn the original. No other references to the argument `payment` survive, so 

Reviewer: ```suggestion
    and delete the original from the issuer's records. Any references to the old payment outside the issuer still will exist, but if anyone attempts to use the old payment, an error will be thrown.
```

    const newPayment = quatloosIssuer.claim(originalPayment, amountToTransfer);
    ```
- [`issuer.combine(paymentsArray)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-combine-paymentsarray)
  - Combine multiple `payments` into one, returned, `payment`. If any `payment` in
  the array is a promise, the operation proceeds after every `payment`
  resolves. All `payments` in the array are burned on successful completion.
  - ```js
    // create an array of 100 payments of 1 unit each
    const payments = [];
    for (let i = 0; i < 100; i += 1) {
      payments.push(quatloosMint.mintPayment(quatloosAmountMath.make(1)));
    }
    // combinedPayment equals 100
    const combinedPayment = quatloosIssuer.combine(payments);
    ```
- [`issuer.split(payment, paymentAmountA)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-split-payment-paymentamounta)
  - Split one `payment` into two new ones, A and B, returned in
    an array. `paymentAmountA` determines A's value, and whatever is
    left of the original `payment` after subtracting A is B's value. 
    If `paymentAmountA` has a larger `value` than `payment`, it throws an error.
    

>     The `payment`
>     argument is burned. If `payment` is a promise, the operation proceeds after

Reviewer: ```suggestion
    The `payment`
    argument is deleted from the issuer's records. If `payment` is a promise for a payment, the operation proceeds after
```

    the promise resolves. 
  - ```js
    const oldPayment = quatloosMint.mintPayment(quatloosAmountMath.make(20));
    // Results in paymentA = 5 and paymentB = 15 (20 -5)
    const [paymentA, paymentB] = quatloosIssuer.split(oldPayment, amountMath.make(5));
    ```
- [`issuer.splitMany(payment, amountArray)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-splitmany-payment-amountarray)
  - Split `payment` into multiple `payments`, returned as an array the
    same length as `amountArray` and with its `payments` having the
    same values as specified for `amountArray`'s elements. If `payment`

>     is a promise, the operation proceeds after it resolves. If

Reviewer: ```suggestion
    is a promise for a payment, the operation proceeds after it resolves. If
```

    the `payment` value is not equal to the sum of `amountArray`'s

>     values, the operation fails. On success, the original `payment` is burned.

Reviewer: ```suggestion
    values, the operation fails. On success, the original `payment` is deleted from the issuer's records.
```

  - ```js
    const oldPayment = quatloosMint.mintPayment(quatloosAmountMath.make(100));
    const goodAmounts = Array(10).fill(quatloosAmountMath.make(10));
    // Results in an array of 10 payments, each with value 10.
    const arrayOfNewPayments = quatloosIssuer.splitMany(oldPayment, goodAmounts);
    ```
- [`issuer.isLive(payment)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-islive-payment)

>   - Returns `true` if `payment` has value. If `payment` is a promise,

Reviewer: ```suggestion
  - Returns `true` if `payment` has value. If `payment` is a promise for a payment,
```

    const quatloos123 = quatloosAmountMath.make(123);
    // Deposit a payment for 123 quatloos into the purse. Ensure that this is what you expect.
    quatloosPurse.deposit(quatloosPayment, quatloos123);
    const secondPayment = quatloosint.mintPayment(quatloosmountMath.make(100));
    // Throws error
    quatloosPurse.deposit(secondPayment, quatloos123);
    ```
- [`purse.makeDepositFacet()`](https://agoric.com/documentation/ertp/api/purse.html#purse-makedepositfacet)
  - Creates a deposit-only facet on the `purse` that can be given to
    other parties that lets them deposit a `payment` (but not
    withdraw) into the `purse`.
  - ```js
    const depositOnlyFacet = purse.makeDepositFacet();
    // Give depositOnlyFacet to someone else. 
    // They can pass a payment that will be deposited:
    depositOnlyFacet.receive(payment);
    ```
## `purse` and `payment` example

The following code creates a new `purse` for the `quatloos` brand, deposits
10 Quatloos into the `purse`, withdraws 3 Quatloos from the `purse` into a
`payment`, and finally returns an `amount` describing what's currently in the `purse`, 7 Quatloos.

```js
// Create a purse with a balance of 10 Quatloos

> const { issuer: quatloosIssuer, mint: quatloosMint } = makeIssuerKit('quatloos');

Reviewer: should we get the amountMath from here too?

const quatloosPurse = quatloosIssuer.makeEmptyPurse();
const quatloosPayment = quatloosMint.mintPayment(quatloosAmountMath.make(10));

> const quatloos10 = quatloosAmountMath.make(10);

Reviewer: why do we make quatloos10 twice, once in line 302 and another in line 303? Can we move line 303 up one and use it instead?

Author: Was trying to show how the optAmount would successfully work with it and the deposited amount being two different amount records with the same value and brand values. Is there a better/different way to do that or is it just not needed?

Reviewer: we could do the following and it would shorter and more in line with how people code:

```
const quatloos10 = quatloosAmountMath.make(10);
const quatloosPayment = quatloosMint.mintPayment(quatloos10);

```
